From edf194863fa874483ad5bc6d9ef3c6fb72f916b8 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:37:14 -0400
Subject: [scsi] don't reference freed command in scsi_prep_return

Message-id: <1402670238-13916-147-git-send-email-snitzer@redhat.com>
Patchwork-id: 83868
O-Subject: [RHEL7.1 PATCH 146/150] scsi: don't reference freed command in scsi_prep_return
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

BZ: 1105204

Upstream commit 68c03d9193f55dad93036f439b94912c5003a173
Author: Christoph Hellwig <hch@lst.de>
Date:   Tue Apr 15 12:24:56 2014 +0200

    [SCSI] don't reference freed command in scsi_prep_return

    Patch

    commit 0479633686d370303e3430256ace4bd5f7f138dc
    Author: Christoph Hellwig <hch@infradead.org>
    Date:   Thu Feb 20 14:20:55 2014 -0800

        [SCSI] do not manipulate device reference counts in scsi_get/put_command

    Introduced a use after free:I in the kill case of scsi_prep_return we have to
    release our device reference, but we do this trying to reference the just
    freed command.  Use the local sdev pointer instead.

    Fixes: 0479633686d370303e3430256ace4bd5f7f138dc
    Reported-by: Joe Lawrence <joe.lawrence@stratus.com>
    Signed-off-by: Christoph Hellwig <hch@lst.de>
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/scsi_lib.c b/drivers/scsi/scsi_lib.c
index a512af7..e44aecb 100644
--- a/drivers/scsi/scsi_lib.c
+++ b/drivers/scsi/scsi_lib.c
@@ -1297,7 +1297,7 @@ int scsi_prep_return(struct request_queue *q, struct request *req, int ret)
    struct scsi_cmnd *cmd = req->special;
    scsi_release_buffers(cmd);
    scsi_put_command(cmd);
-   put_device(&cmd->device->sdev_gendev);
+   put_device(&sdev->sdev_gendev);
    req->special = NULL;
   }
   break;
-- 
1.7.1